### PR TITLE
Handle padding in decoder_inputs_id when using generate

### LINKED ISSUE #10478: generate() decoder_input_ids padding
## Environment info

- `transformers` version: 4.4.0.dev0
- Platform: Linux-5.4.0-66-generic-x86_64-with-debian-buster-sid
- Python version: 3.7.9
- PyTorch version (GPU?): 1.7.1 (True)
- Tensorflow version (GPU?): not installed (NA)
- Using GPU in script?: Yes (but unrelevant)
- Using distributed or parallel set-up in script?: No

### Who can help

As it is a generation issue:
- blenderbot, bart, marian, pegasus, encoderdecoder,  t5: @patrickvonplaten, @patil-suraj 

But also may benefit from better documentation:

Documentation: @sgugger

## Information

When using the `generate()` method from `generation_utils` for Bart (although other models will probably behave the same by checking the code), I am using `decoder_input_ids` to inform of the first tokens each sample in the batch should start with. Although not stated in the documentation (see [post in HF forum](https://discuss.huggingface.co/t/generate-continuation-for-seq2seq-models/)), `generate()` can take `decoder_input_ids` as the forward method would. 

This works fine with batch size equal to one or if the `decoder_input_ids` would all have same length and not require padding. However, when padding is involved, the `generate` method does not ignore the padding tokens in order to generate text for each sample in the batch, generating instead after the padding tokens in `decoder_input_ids`. 

## To reproduce

1. Load your favorite bart model for generation.
2. Prepare your `inputs_ids` for the encoder and the `decoder_input_ids` for your decoder, using sequences of different length.
3. Check the generated text.

```
from transformers import AutoConfig, AutoModelForSeq2SeqLM, AutoTokenizer
tokenizer = AutoTokenizer.from_pretrained("sshleifer/distilbart-cnn-6-6")

model = AutoModelForSeq2SeqLM.from_pretrained("sshleifer/distilbart-cnn-6-6")

model.to("cuda:0")
model.eval()
inputs_text = ['Hugging Face is taking its first step into machine translation this week with the release of more than 1,000 models. Researchers trained models using unsupervised learning and the Open Parallel Corpus (OPUS). OPUS is a project undertaken by the University of Helsinki and global partners to gather and open-source a wide variety of language data sets, particularly for low resource languages. Low resource languages are those with less training data than more commonly used languages like English.', 'Hugging Face has announced the close of a $15 million series A funding round led by Lux Capital, with participation from Salesforce chief scientist Richard Socher and OpenAI CTO Greg Brockman, as well as Betaworks and A.Capital.']
decoder_text = ['</s><s>Hugging Face released', '</s><s>Hugging Face closed a']
inputs = tokenizer(inputs_text, return_tensors = 'pt', padding=True)
inputs_decoder = tokenizer(decoder_text, return_tensors = 'pt', padding=True, add_special_tokens = False)
generated_tokens = model.generate(inputs['input_ids'].to(model.device), 
                                  attention_mask=inputs['attention_mask'].to(model.device), 
                                  decoder_input_ids = inputs_decoder['input_ids'].to(model.device))
decoded_preds = tokenizer.batch_decode(generated_tokens, skip_special_tokens=False)
print('\n'.join(decoded_preds), '\n')
Output:
</s><s>Hugging Face released<pad> models using unsupervised learning and the Open Parallel Corpus (OPUS) Low resource languages are those with less training data than more commonly used languages like English. The project was undertaken by the University of Helsinki and global partners to gather and open-source a wide variety of language data sets.</s>
</s><s>Hugging Face closed a $15 million series A funding round led by Lux Capital. Salesforce chief scientist Richard Socher and OpenAI CTO Greg Brockman participated. Betaworks and A.Capital also took part in the round, with participation from Betaworkorks.</s><pad><pad><pad><pad><pad><pad><pad> 
```
In the first generated sentence, the `<pad>` token after released is kept, and would be worse if there was a higher length difference between the decoder inputs.

## Expected behavior

If one gives as `decoder_input_ids` same length sentences, for instance, we remove the a from "Hugging Face closed a":
```
decoder_text = ['</s><s>Hugging Face released', '</s><s>Hugging Face closed']
inputs = tokenizer(inputs_text, return_tensors = 'pt', padding=True)
inputs_decoder = tokenizer(decoder_text, return_tensors = 'pt', padding=True, add_special_tokens = False)
generated_tokens = model.generate(inputs['input_ids'].to(model.device), 
                                  attention_mask=inputs['attention_mask'].to(model.device), 
                                  decoder_input_ids = inputs_decoder['input_ids'].to(model.device))
decoded_preds = tokenizer.batch_decode(generated_tokens, skip_special_tokens=False)
print('\n'.join(decoded_preds), '\n')
Output:
</s><s>Hugging Face released its first step into machine translation this week. Researchers trained models using unsupervised learning and the Open Parallel Corpus (OPUS) OPUS is a project undertaken by the University of Helsinki and global partners. Low resource languages are those with less training data than more commonly used languages like English.</s>
</s><s>Hugging Face closed a $15 million series A funding round led by Lux Capital. Salesforce chief scientist Richard Socher and OpenAI CTO Greg Brockman were involved. Betaworks and A.Capital also participated in the round, which will take place in New York City.</s><pad><pad><pad><pad><pad> 
```
Then the output does not include any `<pad>` token in between the generated text. It would be nice if this would be the same for different length in the decoder_input_ids (ie. ignore the `<pad>` tokens).

### PR DESCRIPTION
**Pull request by LittlePea13**
# What does this PR do?

Fixes #10478 for pytorch version of `generate()` in generate_utils.py

As described in the issue, when `decoder_input_ids` have different lengths and require padding, generation continues after the padding tokens. This PR modifies that behavior so that tokens are generated before the padding for each element in the batch.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

I am adding those who I see more often in the git blame of the file:
@patrickvonplaten, @patil-suraj, @yjernite

This is my first PR, so let me use it to thank everyone involved in this library for all the cool work :)

### CODE REVIEW

**Comment by LittlePea13** [comment]
Some tests failed due to pad_token_id being None. Is this really a possibility for any Transformer model?

**Comment by LittlePea13** [comment]
Added a check for token_pad_id equal to None and tests pass, but it is not very elegant, any feedback will be appreciated.

**Comment by LittlePea13** [comment]
I just noticed that one test, `run_tests_flax`, failed, however no changes are made that should affect that. @patrickvonplaten or @patil-suraj, when you have the time let me know if there's anything else I should be aware of regarding that test.

**Comment by patil-suraj** [comment]
Hi @LittlePea13 

Thanks a lot for the PR. 

I understand the problem but it seems like an edge case and overall I'm not really in favor of supporting this. The philosophy of `generate` is to keep it simple and extensible and not try to cover all use-cases.
We generally try to keep such if/else statements to a minimum. This change will introduce a lot of complexity in the code.

Also with this change, we won't be able to use `use_cache`, which will slow down generation significantly.

One could always just call generate multiple times if the `decoder_input_ids` are of different length. I would rater batch the sentences together which require the  `decoder_input_ids` of the same length and then pass those to `generate` instead of passing  `decoder_input_ids` of different lengths. Which would cover this use-case.

But thanks a lot for your work! It's a good practice to first propose and discuss the solution in the issues before opening a PR. 

What do you think @patrickvonplaten?

**Comment by LittlePea13** [comment]
Hi @patil-suraj thanks for the feedback, I agree that this introduces something specific, and I am not too happy on how it is dealt with for the reasons you point out. However, it seemed useful to have a way to deal with it since it is what one would expect if one inputs `decoder_input_ids` to the `generate` function with some padding.

I opened an issue but was too impatient and opened a PR (sorry about that), basically because I needed this for my own work and coded it anyways.

Perhaps this doesn't belong here, but in any case I feel like including extra documentation about `decoder_input_ids` [here](https://huggingface.co/transformers/main_classes/model.html?highlight=beam%20search#transformers.generation_utils.GenerationMixin.generate) would be beneficial, maybe explaining this behavior (ie. they have to be of the same length).

**Comment by patrickvonplaten** [comment]
Hey @LittlePea13, 

Thanks for raising awareness for your problem and thanks for opening a PR! I agree with @patil-suraj here and would prefer to not include such specific code in `generate()`. 

In general the philosophy for more specific use cases of `generate()` is to directly use the "sub"-generate methods, such as `sample()`, `greedy_search()`, and `beam_search()` as explained here:  https://discuss.huggingface.co/t/big-generate-refactor/1857

I think in your use case, we could do a similar trick that what was done for GPT2 for batched inference:
https://discuss.huggingface.co/t/batch-generation-with-gpt2/1517

This means that instead of passing `["This is <PAD> <PAD>", "This is a sentence"]` as `decoder_input_ids` you could pass `["<PAD> <PAD> This is", "This is a sentence"]` to make `generate()` work. Could you try this out? Also, I'd recommend to directly use `beam_search()` instead of `generate()` in your example. If it doesn't work feel free to post on the forum: https://discuss.huggingface.co/ and tag me - I'll try to help you make it work then :-)

**Comment by patrickvonplaten** [comment]
I think a well-written forum post would also be a great way of documenting this behavior. However I do think, it's a bit too specific for the general docs of `generate()` since they don't even include `decoder_input_ids` as an input argument to the function.

**Comment by patrickvonplaten** [comment]
@patil-suraj and I will keep an eye out if more people run into this problem! Thanks a lot for bringing it up in any case :-)

**Comment by LittlePea13** [comment]
Thanks both for the review. Indeed, I didn't think of just moving the padding to the left, much more elegant. I tried it out and it works but it produces different outputs than without padding.

The issue here is that by using directly the different "sub"-generate methods is not possible to apply the same changes, so if one wants to have the same results as if there was no difference in the sentences lengths they would still need to do a similar tweak as the one here on each method.

But this is very narrow, I don't even know if it affects performance in my case when compared to moving padding to the left. I am closing this and in case someone has a similar issue you can just refer to the changes here. 

Cheers!
